net/fetcher.go

```diff
@@ -0,0 +1,73 @@
+package net
+
+import (
+	"context"
+
+	"gx/ipfs/QmR8BauakNcBa3RbE4nbQu76PDiJgoQgz8AJdhJuiU4TAw/go-cid"
+	"gx/ipfs/QmVmDhyTTUcQXFD1rRQ64fGLMSAoaQvNH3hwuaCFAPq2hy/errors"
+	"gx/ipfs/QmWoXtvgC8inqFkAATB7cp2Dax7XBi9VDvSg9RCCZufmRk/go-block-format"
+	"gx/ipfs/QmXixGGfd98hN2dA5YiPHWANY3sjmHfZBQk3mLiQUo6NLJ/go-bitswap"
+	bserv "gx/ipfs/QmZsGVGCqMCNzHLNMB6q4F6yyvomqf1VxwhJwSfgo1NGaF/go-blockservice"
+	logging "gx/ipfs/QmbkT7eMTyXfpeyB3ZMxxcxg7XH8t6uXp49jqzz4HB7BGF/go-log"
+
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+var logFetcher = logging.Logger("net.fetcher")
+
+// Fetcher is used to fetch data over the network.  It is implemented with
+// a networked blockservice and a persistent bitswap session.
+type Fetcher struct {
+	// blockService is a network enabled blockstore.
+	blockService bserv.BlockService
+
+	// session is a bitswap session that enables efficient transfer.
+	session *bserv.Session
```

Comment:
it is always so weird to me to see us keep a handle on a thingy, and then also on an underlying thingy. its like fake layering. these stats have very little to do with fetching so if you wanted you could just have a logstats function that is passed the underlying exhange in the caller, and not have the fetcher implement logstats. i guess we want to get this in, so maybe not now, but look for this kind of thing in the future ie single responsibility principle.

Reply:
cc @whyrusleeping I could use your input as I inherited this from your code.

Reply:
i'm not sure i grok @phritz's comment relative to what he's commenting on. The session being here is the whole point, and is an optimized 'fetching thingy'. I think having the blockservice here isnt necessary, and that getting the statistics for block sends and receives should be wired through separately (maybe as a command like we do in ipfs)

Reply:
as i was writing that comment I started to understand what phritz is saying. And agree, the blockservice really doesnt belong here.